blog post about the browser edition (#1120)
deltachat/deltachat-pages: _posts/2025-05-20-browser-edition.md

- 🔑 Password Protection
- ⬇️ Download attachments or open in browser (useful for PDF)
- ℹ️ Local Help also works
- 🌐 Language based on Browser

Here is a video that shows the Browser Version in action:

<video controls style="max-width: 100%;" alt="Demo video of the Delta Chat Web version in action"><source src="https://chatmail.at/video/browser-edition-blogpost-demo.mp4" type="video/mp4"></video>

Though it is not standalone, it still needs a server component because chatmail core[^1] can not yet be compiled into a web assembly to run fully in the browser, but more on that later.

### Why we made a Browser Edition

So why did we make it? We had three reasons:
1. As proof that the delta chat desktop react-based user interface is not dependent on electron anymore (an important stepping stone to the tauri[^3] version).
2. To help us to improve and develop the web-based user interface by giving us access to the development tools and add-ons on all browsers.
3. To bring back automated integration testing for Delta Chat desktop.

### Possible future use cases {#future-usecases}

There are many more ways this web version could be used besides the ones already mentioned:

> -  "delta chat web" - run core on an Android device, connect to it from a computer over a local network for a WhatsApp web-like experience.

Author: won't work on iOS, but well we get to that detail if we ever go down that path

> - delta chat as a service (for example, a company could host instances for all their employees).
> - It could be a way to port Delta Chat to special operating systems[^2] that have a browser and rust support, but have no support for electron or tauri[^3].
> - run delta chat web on your Raspberry Pi / home server connect from your devices

Author: ```suggestion
- Delta chat as a service (for example, a company could host instances for all their employees).
- It could be a way to port Delta Chat to special operating systems[^2] that have a browser and rust support, but have no support for electron or tauri[^3].
- Run delta chat web on your Raspberry Pi / home server connect from your devices
```

	- treefit already made a [plugin for running it on home assistant](https://codeberg.org/treefit/deltachat-homeassistant-addon)

<figure>
    <img alt="Delta Chat Web inside of Home Assistant" src="../assets/blog/2025-05-20-browser-edition/browser-edition-in-home-assistant.png" style="max-width: 100%" />
    <figcaption>Delta Chat Web inside of Home Assistant</figcaption>
</figure>

### Technical Details for the Interested

To make the Web UI of delta chat desktop independent of electron, we needed to make our code more modular:

- We switched to the JSON-RPC API, which we explored in detail in a [previous blog post](https://delta.chat/en/2025-02-11-why-jsonrpc-bindings-exist).

> - We created a new `Runtime` interface and moved all direct calls to electron into a `ReuntimeElectron` class that implements this interface.

Reviewer: Spelling mistake of "ReuntimeElectron" -> RuntimeElectron

Author: ```suggestion
- We created a new `Runtime` interface and moved all direct calls to electron into a `RuntimeElectron` class that implements this interface.
```

    <p>The current approach has the following caveats that you should to keep in mind</p>
    <ol>
        <li>
            If you host the server component on a VPS, then the VPS will become the place where the messages are decrypted, so you break the premise of the end-to-end encryption: "end device to end device".
        </li>
        <li>
            You need to host the server component for each user, so you need to build management software if you want to use this for a SaaS project/product.
        </li>
        <li>
            Currently, only one client can connect to the chatmail core at a time, because there is only a single event queue. <br /> But there are ideas to change this in the future to allow for more concurrent event listeners. (If you would connect multiple clients right now, then they would steal events from each other.).
        </li>
    </ol>
    <p>
    Also, not all features are implemented yet in the browser version, mainly maps/location-streaming, webxdc mini apps, and viewing HTML emails. But those are not impossible to implement; we just need some considerations about sandboxing.
    </p>
</details>

### What's next?

> While we have no immediate plans for the browser version (besides using it for more e2e testing), there are many ideas on what it could be used for ([Possible future use cases](#future-usecases)).
> 
> Ideally, the web version shouldn't need this local server component. To make that work, we need to compile our core library to wasm. It should be possible since our core library is written in rust, and rust is one of the most popular languages for compiling to web assembly, but there are a few challenges (missing TCP support in browsers, a fast SQLite database, fast file storage, threading, and more).
> If you want to help, here is the forum discussion about it: [What would be needed for a standalone web version without a server component? | Delta Chat Forum](https://support.delta.chat/t/what-would-be-needed-for-a-standalone-web-version-without-a-server-component/3789).

Reviewer 2: ```suggestion
Besides the issues noted, a Web version of Delta Chat that fully satisfies end-to-end encryption guarantees needs the Rust chatmail core library to run in the browser. Rust in general compiles to [WebAssembly (WASM)](https://webassembly.org/). For example, the security-audited [rPGP end-to-end encryption library](https://github.com/rpgp/rpgp) is implemented fully in Rust and  is continuously tested with WebAssembly targets. However, there are some key challenges to address for a "standalone" Web version: 

 - find a solution for both database storage (currently chatmail core uses sqlite as an embedded C-library) and arrange fast file storage for media files, avatars etc.   

- find a solution to Browsers being unable to perform SMTP or IMAP network protocols; this could involve [[Chatmail relays](https://chatmail.at/relays) offering a minimal HTTP interface

- support webxdc realtime P2P networking and support running [in-chat multiplayer Quake3 Arena realtime gameplay](https://chaos.social/@delta/114517181096683376);  Our friends at [Iroh](https://iroh.computer) are themselves working on Web-versions 

- explore how well the chatmail core async Rust code can run in WebAssembly; this could involve a lot of refactoring 

See the [Web version topic in the Delta Chat Forum](https://support.delta.chat/t/what-would-be-needed-for-a-standalone-web-version-without-a-server-component/3789) for further discussion. 

If you are not really into programming, it might be hard to understand how hard these issues are. But no worries, even if you are a programmer, or even with ourselves as matter experts, it's hard to predict :) 

## Invitation to DIFF June 7-17th in Freiburg

The best way to discuss with many of us is by making it to an in-person community gathering. This crazy year of 2025 around we are [inviting to the DIFF gathering](https://delta.chat/en/2025-05-12-diff-invitation) starting in just a few weeks.  

## Donations needed

While some of our work is funded through public bodies, a lot of it is not.  Please consider [sending a monetary contribution](https://delta.chat/en/donate) if you can and appreciate our efforts and want to enable more. Thanks! 
```

 - find a solution for both database storage (currently chatmail core uses sqlite as an embedded C-library) and arrange fast file storage for media files, avatars etc.   

- find a solution to Browsers being unable to perform SMTP or IMAP network protocols; this could involve [[Chatmail relays](https://chatmail.at/relays) offering a minimal HTTP/Websocket interface to bridge the gap.

- support webxdc realtime P2P networking and support running [in-chat multiplayer Quake3 Arena realtime gameplay](https://chaos.social/@delta/114517181096683376);  Our friends at [Iroh](https://iroh.computer) are themselves working on Web-versions 

- explore how well the chatmail core async Rust code can run in WebAssembly; this could involve a lot of refactoring.

See the [Web version topic in the Delta Chat Forum](https://support.delta.chat/t/what-would-be-needed-for-a-standalone-web-version-without-a-server-component/3789) for further discussion. 

If you are not really into programming, it might be hard to understand how hard these issues are.

> But no worries, even if you are a programmer, or even with ourselves as matter experts, it's hard to predict :) 

Author: this sentence is a bit hard to read, we should simplify it somehow.
